boa_engine/src/host_defined.rs

```diff
@@ -1,111 +1,145 @@
 use std::any::TypeId;
 
-use boa_gc::{GcRef, GcRefCell, GcRefMut};
 use boa_macros::{Finalize, Trace};
-use rustc_hash::FxHashMap;
+use hashbrown::hash_map::HashMap;
 
 use crate::object::NativeObject;
 
-/// Map used to store the host defined objects.
-#[doc(hidden)]
-type HostDefinedMap = FxHashMap<TypeId, Box<dyn NativeObject>>;
-
 /// This represents a `ECMASCript` specification \[`HostDefined`\] field.
 ///
 /// This allows storing types which are mapped by their [`TypeId`].
 #[derive(Default, Trace, Finalize)]
 #[allow(missing_debug_implementations)]
 pub struct HostDefined {
-    state: GcRefCell<HostDefinedMap>,
+    types: HashMap<TypeId, Box<dyn NativeObject>>,
+}
+
+#[allow(unsafe_op_in_unsafe_fn)]
```

Comment:
We turned this lint on because rustc is planning to warn on all usages of unsafe operations outside `unsafe` blocks, even if they're inside an `unsafe` function. We should do that and document the safety invariants of the `unsafe` block instead.

Reply:
I documented the internal invariants of `HostDefined` and a safety comment to `downcast_boxed_native_object_unchecked`.

Reply:
Nice! You can just remove the lint and add an `unsafe` block after the `SAFETY` comment now. That should fix it. Something like:
```rust
let raw: *mut dyn NativeObject = Box::into_raw(obj);

// SAFETY: safety comment
unsafe {
    Box::from_raw(raw.cast::<T>())
}
```

Reply:
Done :)